src/Parser.php

```diff
@@ -361,24 +338,15 @@ private function filterBodyNodes(Node $node, bool $nested = false)
             return $node;
         }
 
-        // to be removed completely in Twig 3.0
-        if (!$nested && $node instanceof SpacelessNode) {
-            @trigger_error(sprintf('Using the spaceless tag at the root level of a child template in "%s" at line %d is deprecated since Twig 2.5.0 and will become a syntax error in 3.0.', $this->stream->getSourceContext()->getName(), $node->getTemplateLine()), E_USER_DEPRECATED);
-        }
-
         // "block" tags that are not captured (see above) are only used for defining
         // the content of the block. In such a case, nesting it does not work as
         // expected as the definition is not part of the default template code flow.
-        if ($nested && ($node instanceof BlockReferenceNode || $node instanceof \Twig_Node_BlockReference)) {
-            //throw new SyntaxError('A block definition cannot be nested under non-capturing nodes.', $node->getTemplateLine(), $this->stream->getSourceContext());
-            @trigger_error(sprintf('Nesting a block definition under a non-capturing node in "%s" at line %d is deprecated since Twig 2.5.0 and will become a syntax error in 3.0.', $this->stream->getSourceContext()->getName(), $node->getTemplateLine()), E_USER_DEPRECATED);
-
-            return;
+        if ($nested && $node instanceof BlockReferenceNode) {
+            throw new SyntaxError('A block definition cannot be nested under non-capturing nodes.', $node->getTemplateLine(), $this->stream->getSourceContext());
         }
 
-        // the "&& !$node instanceof SpacelessNode" part of the condition must be removed in 3.0
-        if ($node instanceof NodeOutputInterface && !$node instanceof SpacelessNode) {
-            return;
+        if ($node instanceof NodeOutputInterface) {
+            return null;
         }
 
         // here, $nested means "being at the root level of a child template"
```

Comment:
the `'Twig_Node' !== \get_class($node)` part of the next condition must be removed too

Reply:
see #2972